state.c

```diff
@@ -0,0 +1,337 @@
+
+    signald_parse_account_list(sa, json_object_get_array_member(data, "accounts"));
+
+    return TRUE;
+}
+
+void
+signald_init_state_machine()
+{
+    if (entrypoint != NULL) {
+        current = entrypoint;
+        return;
+    }
+
+    SignaldState *version;
+    SignaldState *subscribed, *linking, *linked;
+    SignaldState *user, *user_not_registered, *unexpected_error;
+    SignaldState *contacts;
+    SignaldState *running;
+
+    /*
+     * Initialization portion of the state machine.
+     *
+     * From "start" we can get the version and then go to linking or subscribing.
+     *
+     * At the end of both paths we end up in "subscribed".
+     */
+    entrypoint = signald_new_state(NULL, "start", NULL, NULL, NULL);
+
+    version = signald_new_state(entrypoint, "got version, waiting", "version", signald_received_version, NULL);
+
+    /*
+     * Easy case, already subscribed.  Just get the message and move on to
+     * checking the protocol version.
+     */
+    subscribed = signald_new_state(version, "subscribe successful, check protocol", "subscribed", signald_subscribed, signald_check_proto_version);
+
+    /*
+     * The fun case.  Request the linking URI, wait for success (or error),
+     * then if successful, subscribe.
+     */
+    linking = signald_new_state(version, "got linking uri, waiting", "linking_uri", signald_linking_uri, NULL);
+    linked = signald_new_state(linking, "linking successful, subscribing", "linking_successful", signald_linking_successful, signald_subscribe_after_link);
+    signald_add_transition(linked, "subscribed", subscribed, signald_subscribed, signald_check_proto_version);
+
+    signald_new_state(linking, "error linking", "linking_error", signald_linking_error, NULL);
+
+    /*
+     * Alright, we're subscribed and we've initiated messaging to check the
+     * protocol version.
+     *
+     * What we get back determines the protocol version, and then we request
+     * the contact list.
+     */
+    // If we get a user back, we have the new protocol
+    user = signald_new_state(subscribed, "new protocol detected, get contacts", "user", signald_new_protocol, signald_initialize_contacts);
+    // If we get no user back, we have the new protocol
+    user_not_registered = signald_new_state(subscribed, "new protocol detected, get contacts", "user_not_registered", signald_new_protocol, signald_initialize_contacts);
+    // If we get an error, we have the old protocol
+    unexpected_error = signald_new_state(subscribed, "old protocol detected, get contacts", "unexpected_error", signald_old_protocol, signald_initialize_contacts);
+
+    /*
+     * Once the contact list comes back, we request the group list.
+     */
+    contacts = signald_new_state(user, "got contacts, get groups", "contact_list", signald_received_contact_list, signald_get_groups);
+    signald_add_transition(user_not_registered, "get_contacts", contacts, signald_received_contact_list, signald_get_groups);
```

Comment:
Whoops, these two lines are broken.  Should be "contact_list" not "get_contacts".